src/java.base/share/native/libjli/java.c

```diff
@@ -534,6 +545,260 @@ invokeInstanceMainWithoutArgs(JNIEnv *env, jclass mainClass) {
+                if (NULL == unprivilegedMode) {
+                    JLI_ReportErrorMessage("Failed to allocate memory for the CRIU unprivileged mode.");
+                    *error = MEMORY_ALLOCATION_ERROR;
+                    return NULL;
+                }
+                return unprivilegedMode;
+            } else {
+                JLI_ReportErrorMessage(
+                    "The value %s of the command line option %s is not expected.",
+                    unprivilegedModePropertyValue,
+                    unprivilegedModePropertyName);
+                freeCommandLineOptionValue(unprivilegedModePropertyValue);
+                *error = INVALID_OPTION_VALUE;
+                return NULL;
+            }
+    }
+    return NULL;
+}
+
+static char *getLogFile(int argc, char **argv, int *error) {
+    const char *logFilePropertyName = "-Dopenj9.internal.criu.logFile";
+    char *logFilePropertyValue = getCommandLineOptionValue(logFilePropertyName, argc, argv, error);
+    if (!*error) {
+        if (logFilePropertyValue) {
+            int logFileLength = snprintf(NULL, 0, "--log-file=%s", logFilePropertyValue) + 1;
+            if (0 > logFileLength) {
+                JLI_ReportErrorMessage("Failed to calculate the length of the CRIU log file string.");
+                freeCommandLineOptionValue(logFilePropertyValue);
+                *error = MEMORY_ALLOCATION_ERROR;
+                return NULL;
+            }
+            char *logFile = (char *)JLI_MemAlloc(logFileLength);
+            if (NULL == logFile) {
+                JLI_ReportErrorMessage("Failed to allocate memory for the CRIU log file option.");
+                *error = MEMORY_ALLOCATION_ERROR;
+                return NULL;
+            }
+            snprintf(logFile, logFileLength, "--log-file=%s", logFilePropertyValue);
+            freeCommandLineOptionValue(logFilePropertyValue);
+            return logFile;
+        } else {
+            JLI_ReportErrorMessage("The value of the command line option %s was not found.", logFilePropertyName);
+            *error = OPTION_VALUE_NOT_FOUND;
+            return NULL;
+        }
+    }
+    return NULL;
+}
+
+static int restoreFromCheckpoint(char *checkpointDir, char *logLevel, char *unprivilegedMode, char *logFile) {
+    int argc = 6;
+    char *argv[9] = { NULL };
+    argv[0] = "criu";
+    argv[1] = "restore";
+    argv[2] = "-D";
+    argv[3] = checkpointDir;
+    argv[4] = logLevel;
+    argv[5] = "--shell-job";
+    if (unprivilegedMode) {
+        argv[argc++] = unprivilegedMode;
+    }
+    if (logFile) {
+        argv[argc++] = logFile;
+    }
+    argv[argc] = NULL;
+    if (-1 == execvp(argv[0], argv)) {
+        return -1;
+    }
+    return 0;
+}
+
+static void handleCRaCRestore(int argc, char **argv) {
+    int error = 0;
+    char *checkpointDir = getCheckpointDir(argc, argv, &error);
+    if (OPTION_NOT_FOUND == error) {
+        freeCommandLineOptionValue(checkpointDir);
```

Comment:
Is it possible to return a non-null `checkpointDir` required to be freed when the `error` is `OPTION_NOT_FOUND`?

Reply:
Is `freeCommandLineOptionValue(checkpointDir)` still needed if that's the case?

Reply:
> It is not possible to return a non-null checkpointDir required to be freed when the error is OPTION_NOT_FOUND.

> Is freeCommandLineOptionValue(checkpointDir) still needed if that's the case?

`freeCommandLineOptionValue(checkpointDir)` is not needed if that's the case, but it seems like we could keep it for consistency or future modifications to `getCheckpointDir` that might return non-NULL values even when `OPTION_NOT_FOUND` occurs.